File: ruby/src/sorts/bubble_sort.rb
require_relative "../util/util"

module BubbleSort
    def BubbleSort.sort(arr)
        # copy input array
        out = arr.map { |x| x }

        while !Util.is_sorted(out)
            # go through the array until it's sorted
            for i in (1...out.length)
                # if these values are out of order
                if out[i] < out[i - 1]
                    # swap the values
                    out[i], out[i - 1] = out[i - 1], out[i]
                end
            end
        end

        out
    end
end

File: go/src/sorts/bubblesort.go
package sorts

import "../helper"

func BubbleSort(arr []int) (sorted []int) {
  sorted = make([]int, len(arr))
  copy(sorted, arr)

  for !helper.IsSorted(sorted) {
    for i := 0; i < len(sorted) - 1; i++ {
      if sorted[i] > sorted[i + 1] {
        sorted = helper.Swap(sorted, i, i + 1)
      }
    }
  }

  return
}
